### PR TITLE
ARROW-10148: [Rust] Improved rust/lib.rs that is shown in docs.rs

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This is a revamp of the landing page of Arrow's crate in docs.rs.

* Moved most documentation from README to `src/lib.rs`
* expanded documentation with implemented functionality, including links to each part in the project
* added a minimal example demonstrating basic functionality

How it looks like:

Screenshot 2020-09-30 at 21 27 10

[current version](https://docs.rs/arrow/1.0.1/arrow/) for comparison

I moved docs from the `README` to this page, so that we do not have to maintain it in two places, and added a link in the README to the `src/lib.rc` to refer for details.

### CODE REVIEW

File: rust/arrow/src/ipc/gen/SparseTensor.rs [hunk 1]
```diff
@@ -389,7 +389,7 @@ impl<'a> SparseMatrixIndexCSX<'a> {
     }
     /// indptrBuffer stores the location and size of indptr array that
     /// represents the range of the rows.
-    /// The i-th row spans from indptr[i] to indptr[i+1] in the data.
+    /// The i-th row spans from `indptr[i]` to `indptr[i+1]` in the data.
```
Comment (alamb) [inline rust/arrow/src/ipc/gen/SparseTensor.rs:392]: I was seeing `cargo doc` warnings about these lines just this morning. 👍  for fixing them

File: rust/arrow/src/lib.rs [hunk 2]
```diff
@@ -18,8 +18,109 @@
...
+//! Every [`Array`](array::Array) in this crate has an associated [`DataType`](datatypes::DataType),
+//! that specifies how its data is layed in memory and represented.
+//! Thus, a central enum of this crate is [`DataType`](datatypes::DataType), that contains the set of valid
+//! DataTypes in the specification. For example, [`DataType::Utf8`](datatypes::DataType::Utf8).
+//!
+//! ## Array
+//!
+//! The central trait of this package is the dynamically-typed [`Array`](array::Array) that
+//! represents a fixed-sized, immutable, Send + Sync Array of nullable elements. An example of such an array is [`UInt32Array`](array::UInt32Array).
+//! One way to think about an arrow [`Array`](array::Array) isa `Arc<[Option<T>; len]>` where T can be anything ranging from an integer to a string, or even
```
Comment (alamb) [inline rust/arrow/src/lib.rs:32]:
```suggestion
//! One way to think about an arrow [`Array`](array::Array) is a `Arc<[Option<T>; len]>` where T can be anything ranging from an integer to a string, or even
```

File: rust/arrow/src/lib.rs [hunk 3]
```diff
@@ -18,8 +18,109 @@
...
+//! use arrow::array::{UInt32Array, ArrayRef};
+//! # let array = UInt32Array::from(vec![Some(1), None, Some(3)]);
+//! let array: ArrayRef = Arc::new(array);
+//! assert_eq!(array.len(), 3);
+//! // array.value() is not available in the dynamcally-typed version
+//! assert_eq!(array.is_null(1), true);
+//! assert_eq!(array.data_type(), &DataType::UInt32);
+//! ```
+//!
+//! to downcast, use `as_any()`:
```
Comment (alamb) [inline rust/arrow/src/lib.rs:60]: 👍

File: rust/arrow/src/lib.rs [hunk 4]
```diff
@@ -18,8 +18,109 @@
...
+//! Data in [`Array`](array::Array) is stored in [`ArrayData`](array::data::ArrayData), that in turn
+//! is a collection of other [`ArrayData`](array::data::ArrayData) and [`Buffers`](buffer::Buffer).
+//! [`Buffers`](buffer::Buffer) is the central struct that array implementations use keep allocated memory and pointers.
+//! The [`MutableBuffer`](buffer::MutableBuffer) is the mutable counter-part of[`Buffer`](buffer::Buffer).
+//! These are the lowest abstractions of this crate, and are used throughout the crate to
+//! efficiently allocate, write, read and deallocate memory.
+//!
+//! ## Field, Schema and RecordBatch
+//!
+//! [`Field`](datatypes::Field) is a struct that contains an arrays' metadata (datatype and whether its values
```
Comment (alamb) [inline rust/arrow/src/lib.rs:82]:
```suggestion
//! [`Field`](datatypes::Field) is a struct that contains an array's metadata (datatype and whether its values
```

File: rust/arrow/src/lib.rs [hunk 5]
```diff
@@ -18,8 +18,109 @@
...
+//! can be null), and a name. [`Schema`](datatypes::Schema) is a vector of fields with optional metadata, and together with
```
Comment (alamb) [inline rust/arrow/src/lib.rs:83]:
```suggestion
//! can be null), and a name. [`Schema`](datatypes::Schema) is a vector of fields with optional metadata. 
```

File: rust/arrow/src/lib.rs [hunk 6]
```diff
@@ -18,8 +18,109 @@
...
+//! Together, they form the basis of a schematic representation of a group of [`Arrays`](array::Array).
+//!
+//! In fact, [`RecordBatch`](record_batch::RecordBatch) is a struct with a [`Schema`](datatypes::Schema) and a vector of
+//! [`Array`](array::Array)s, all with the same `len`. A record batch is the highest order struct that this crate currently offersm
```
Comment (alamb) [inline rust/arrow/src/lib.rs:87]:
```suggestion
//! [`Array`](array::Array)s, all with the same `len`. A record batch is the highest order struct that this crate currently offers
```

File: rust/arrow/src/lib.rs [hunk 7]
```diff
@@ -18,8 +18,109 @@
...
+//! This crate offers many operations (called kernels) to operate on `Array`s, that you can find at [compute::kernels].
+//! It has both vertial and horizontal operations, and some of them have an SIMD implementation.
+//!
+//! ## Status
+//!
+//! This crate has most of the implementation of the arrow specification. Specifically, it supports the following types:
+//!
+//! * All arrow primitive types, such as [`Int32Array`](array::UInt8Array), [`BooleanArray`](array::BooleanArray) and [`Float64Array`](array::Float64Array).
+//! * All arrow variable length types, such as [`StringArray`](array::StringArray) and [`BinaryArray`](array::BinaryArray)
+//! * All composite types such as [`StructArray`](array::StructArray) and [`ListArray`](array::ListArray)
```
Comment (alamb) [inline rust/arrow/src/lib.rs:101]:
```suggestion
//! * All composite types such as [`StructArray`](array::StructArray) and [`ListArray`](array::ListArray)
//! * Dictionary types  [`DictionaryArray`](array::DictionaryArray) 

```

**Review by alamb — APPROVED** [review]
Thank you @jorgecarleitao  -- this is a nice change indeed. I left some minor comments but overall I think this is a significant improvement

**Review by andygrove — APPROVED** [review]

**Pull request closed without merging**